Implement tag based documentation

AshesITR opened the pull request:
Closes #888 

So far:

 - [x] Implemented `available_linters(packages = "lintr")` to return a tibble of available linters contained in the packages specified
 - [x] Added tests to make sure `available_linters(packages = ...)` always returns a data frame
 - [x] Added a test to ensure the `default` tag is consistent with `default_linters`
 - [x] Described the format of `inst/lintr/linters.csv` in a Package Authors section of `?available_linters`
 - [x] Made each linter have its own help page with a separate Tags section pointing to its tags
 - [x] Added a list of linters to `?linters`
 - [x] Created new topics `tag_linters` for all tags, each listing all linters tagged with the tag, plus a small description
 - [x] Added a list of tags to all linter help pages
 - [x] Added a list of tags to `?linters`
 - [x] Update NEWS when changes are final

AshesITR commented:
A test that might be worth adding would be to check that all exported functions with the `_linter` suffix appear in `linters.csv`.
I don't know whether it would be wise to self-test the packages own exports.

This test would pass currently:
```
expect_setequal(
  grep("_linter$", ls(asNamespace("lintr")), value = TRUE),
  c(available_linters()[["linter"]], "make_object_linter", "make_path_linter")
)
```

jimhester commented:
We may want to use this PR as an opportunity to switch to markdown style roxygen documentation in lintr. e.g. run `usethis::use_roxygen_md()`

AshesITR commented:
We might want to also `@seealso` the linters help page for all linters and all tags, to improve discoverability.

MichaelChirico commented:
The PR is quite large... it will be easier to review if subdivided. at a glance each of your bullets could be their own PR (though feel free to rearrange).

if we want to avoid merging to main until it's done, let's have a sequence of PRs that we merge into this branch, then to main, WDYT?

AshesITR commented:
Hmm I'm not so sure this will make it easier to review, since multiple of those points touch nearly every file in the codebase.

If it helps, I can try to write down my refactoring process?

MichaelChirico commented:
I think even so, if I can scroll through all the files looking for just one thing that'll be easier than checking for several things at once

AshesITR commented:
I see. Will try to make a somewhat sensible sequence of PRs into this branch, force-resetting this one to step 1.

MichaelChirico commented:
thanks! appreciate the extra effort

AshesITR commented:
@MichaelChirico I've split up the PR so far into 3 separate steps.
You can find Step 1 here, Steps 2 and 3 as AshesITR/lintr#1 and AshesITR/lintr#2 respectively.

I plan on adding a Step 4 to include the suggestion by @jimhester to switch to md based roxygen, as well as to add `@seealso linters for a full documentation of available linters` to all `?*_linter`, `?*_linters` and `?available_linters` help pages.

MichaelChirico commented:
> @seealso linters for a full documentation of available linters

how does that translate into the .Rd exactly? `\seealso{\code{\link{linters}} for full documentation of available linters}`?

AshesITR commented:
Exactly. See AshesITR/lintr#3.
I opted for a slightly different wording in my first try:

```
\seealso{
\link{linters} for a complete list of linters available in lintr.
}
```

AshesITR commented:
I'm done with Step 4 now.
My suggestion would be to review (comment) in order the PRs
#908, AshesITR/lintr#1, AshesITR/lintr#2, AshesITR/lintr#3.

In case of changes in a step, I would rebase the following PRs on the altered version of the previous step, so in the end we could merge all steps into #908 and finally write up a NEWS bullet and merge #908 into master.

MichaelChirico commented:
OK thanks. I take it that's why this is still marked as draft -- kept as draft until the other PRs are merged here. Starting to review now.

MichaelChirico commented on R/linter_tags.R:
```diff
@@ -0,0 +1,186 @@
+#'   'tags'.}
+#' }
+#'
+#' Tags should be snake_case.
```
this doesn't really matter right?

AshesITR replied:
They must not contain spaces, but a "should" recommendation to stick to what we use seemed useful to me.

MichaelChirico replied:
to me "should" read as a requirement. maybe

> In accordance with the style guide, we recommend tags be in snake_case.

conveys it better?

MichaelChirico commented on R/linter_tags.R:
```diff
@@ -0,0 +1,186 @@
+#' A data frame with columns 'linter', 'package' and 'tags':
+#'
+#' \describe{
+#' \item{linter}{A character column naming the function associated to the linter.}
```
associated to --> associated with

AshesITR replied:
fixed

MichaelChirico commented on R/linter_tags.R:
```diff
@@ -0,0 +1,186 @@
+#'
+#' \describe{
+#' \item{linter}{A character column naming the function associated to the linter.}
+#' \item{package}{A characetr column containing the name of the package providing the linter.}
```
typo: characetr

AshesITR replied:
fixed

MichaelChirico commented on R/linter_tags.R:
```diff
@@ -0,0 +1,186 @@
+#' \describe{
+#' \item{linter}{A character column naming the function associated to the linter.}
+#' \item{package}{A characetr column containing the name of the package providing the linter.}
+#' \item{tags}{A list column containing tags associated to the linter.}
```
associated to --> associated with

AshesITR replied:
fixed

MichaelChirico commented on R/linter_tags.R:
```diff
@@ -0,0 +1,186 @@
+  # Make sure we always return a valid data frame
+  # data.frame(...) and as.data.frame(...) don't handle zero-row list-columns properly, so we need to manually create
+  # the structure.
+  empty_linters <- structure(list(
```
I try and avoid `structure()` wherever possible, I think we can make do with `empty_linters <- data.frame(linter =, package =); empty_linters$tags <- list()`

AshesITR replied:
changed

MichaelChirico commented on R/linter_tags.R:
```diff
@@ -0,0 +1,186 @@
+
+  # Check that the csv file contains two character columns, named 'linter' and 'tags'.
+  # Otherwise, fallback to an empty data frame.
+  if (!ncol(available) == 2L) {
```
I think we can be more flexible here, maybe downstream will eventually add their own columns. We can just search for `linter`/`tags` in the columns and return those columns, only error if those two aren't present.

AshesITR replied:
changed to requiring those two columns

MichaelChirico commented on R/linter_tags.R:
```diff
@@ -0,0 +1,186 @@
+    return(empty_linters)
+  }
+
+  structure(list(
```
ditto here for avoiding `structure()`. We could also use the `data.frame(a = 1, b = I(list(2)))` approach but `AsIs` columns have caused nothing but headaches for me in general.

AshesITR replied:
changed in the same way as above

MichaelChirico commented on R/linter_tags.R:
```diff
@@ -0,0 +1,186 @@
+  c(
+    "\\section{Tags}{",
+    if (length(tags)) {
+      paste(
```
i prefer using `toString()` since it's there already

AshesITR replied:
changed

MichaelChirico commented on R/linter_tags.R:
```diff
@@ -0,0 +1,186 @@
+      c(
+        paste0("The following linters are tagged with '", tag_name, "':"),
+        "\\itemize{",
+        vapply(tagged, function(linter) {
```
why do we need `vapply()` here?

AshesITR replied:
changed

MichaelChirico commented on R/linter_tags.R:
```diff
@@ -0,0 +1,186 @@
+    "\\section{Tags}{",
+    if (length(tags)) {
+      paste(
+        vapply(tags, function(tag) {
```
why do we need `vapply()` here?

AshesITR replied:
changed

MichaelChirico commented on R/linter_tags.R:
```diff
@@ -0,0 +1,186 @@
+    "\\itemize{",
+    vapply(linter_names, function(linter_name) {
+      tags <- sort(linters[["tags"]][[match(linter_name, linters[["linter"]])]])
+      paste0("\\item{\\code{\\link{", linter_name, "}} (tags: ", paste(tags, collapse = ", "), ")}")
```
nit: `toString()` here too

AshesITR replied:
changed

MichaelChirico commented on tests/testthat/test-linter_tags.R:
```diff
@@ -0,0 +1,20 @@
+  avail2 <- available_linters(c("lintr", "not-a-package"))
+  empty <- available_linters("not-a-package")
+
+  expect_is(avail, "data.frame")
```
expect_is() is deprecated, use expect_s3_class()

AshesITR replied:
changed

MichaelChirico commented on tests/testthat/test-linter_tags.R:
```diff
@@ -0,0 +1,20 @@
+  expect_is(avail, "data.frame")
+  expect_identical(avail, avail2)
+  expect_equal(nrow(empty), 0L)
+  expect_equal(colnames(avail), c("linter", "package", "tags"))
```
use expect_named() here

AshesITR replied:
changed

MichaelChirico commented on tests/testthat/test-linter_tags.R:
```diff
@@ -0,0 +1,20 @@
+
+  expect_is(avail, "data.frame")
+  expect_identical(avail, avail2)
+  expect_equal(nrow(empty), 0L)
```
After https://github.com/r-lib/lintr/pull/910 we should use `expect_identical()` by default:

> `expect_identical()` is the baseline; `expect_equal()` relaxes the test to ignore small numeric differences.

AshesITR replied:
changed

AshesITR commented:
Thanks @MichaelChirico for your suggestions. I incorporated all of them and rebased the subsequent PRs.
Do you have an idea why the new code suddenly returns the wrong data type in older R versions?
Tempted to revert to the `structure()` based code which was stable across all versions...

MichaelChirico commented:
looks like it was a stringsAsFactors issue

MichaelChirico commented on R/linter_tags.R:
```diff
@@ -0,0 +1,174 @@
+#'
+#' To implement \code{available_linters()} for your package, include a file \code{inst/lintr/linters.csv} in your
+#' package.
+#' The CSV file must contain two columns, 'linter' and 'tags', and be UTF-8 encoded.
```
nit: slight tweak here to convey those columns are required, but other columns are allowed. maybe just remove "two"?

AshesITR replied:
done

MichaelChirico commented on R/linter_tags.R:
```diff
@@ -0,0 +1,174 @@
+#'
+#' \enumerate{
+#' \item{its function name (e.g. \code{"assignment_linter"}) in the column 'linter'.}
+#' \item{space separated tags associated with the linter (e.g. \code{"style consistency default"}) in the column
```
nit: space-separated

AshesITR replied:
fixed

MichaelChirico reviewed (approved):
looks great thanks! I know this approval will get reset, just signalling it's GTG for this part

AshesITR commented:
Thanks for the feedback so far.
Do you want to start reviewing AshesITR/lintr#1 or should I merge that branch into #908 (enlarging the diff)?

MichaelChirico commented:
I'll review the other PR, then we can merge to here. GitHub doesn't make it easy to say "I've reviewed through commit X, show me the diff relative to that" (it's possible but cumbersome)

AshesITR commented:
The oldrel-4 build failure is due to a bug in rmarkdown:
https://github.com/rstudio/rmarkdown/issues/2324

AshesITR commented:
@MichaelChirico, @jimhester PTAL.
I've added a NEWS bullet with multiple sub-bullets. Not sure if that's the optimal way to present the information since describing this PR in a single bullet seems difficult.

AshesITR commented on NEWS.md:
```diff
@@ -79,6 +79,12 @@ function calls. (#850, #851, @renkun-ken)
 * Debugging functions (`browser()`, `debug()`, `debugcall()`, `debugonce()`, `trace()`, `undebug()`, `untrace()`) are now part of the default set of undesirable functions to help prevent them from being committed by mistake. (#876, @michaelchirico)
 * New linter `package_hooks_linter()` runs a series of checks also done by `R CMD check` on the `.onLoad()`, `.onAttach()`, `.Last.lib()` and `.onDetach()` hooks (#882, @MichaelChirico)
 * Improved location information for R parse errors (#894, #892, @renkun-ken and @AshesITR)
+* New tag based documentation pages for linters (#888, @AshesITR)
+  * Each linter has its own help page
+  * `?linters` also links to tag help pages, collecting linters with a similar goal
+  * Each linter can have multiple tags
+  * New function `available_linters()` to list available linters and their tags 
+    This feature is extensible by package authors providing add-on linters for {lintr}.
```
Alternatively, we could add the bullets on the top level and duplicate the issue number?

MichaelChirico replied:
this bullet is pretty vague maybe a See ?available_linters reference?

MichaelChirico commented on NEWS.md:
```diff
@@ -79,6 +79,12 @@ function calls. (#850, #851, @renkun-ken)
 * Debugging functions (`browser()`, `debug()`, `debugcall()`, `debugonce()`, `trace()`, `undebug()`, `untrace()`) are now part of the default set of undesirable functions to help prevent them from being committed by mistake. (#876, @michaelchirico)
 * New linter `package_hooks_linter()` runs a series of checks also done by `R CMD check` on the `.onLoad()`, `.onAttach()`, `.Last.lib()` and `.onDetach()` hooks (#882, @MichaelChirico)
 * Improved location information for R parse errors (#894, #892, @renkun-ken and @AshesITR)
+* New tag based documentation pages for linters (#888, @AshesITR)
```
it would help to motivate this (why as lintr user do I care about this update?). maybe a quick blurb like

> Add new tag-based documentation pages for linters to help organize information surrounding the growing number of linters available

MichaelChirico replied:
Refining:

> To help organize information surrounding the growing number of linters available, we've added metadata "tags" for each linter `(#888)`.

> bullet: each tag gets a help page, e.g. ?x

WDYT

MichaelChirico commented on NEWS.md:
```diff
@@ -79,6 +79,12 @@ function calls. (#850, #851, @renkun-ken)
 * New linter `package_hooks_linter()` runs a series of checks also done by `R CMD check` on the `.onLoad()`, `.onAttach()`, `.Last.lib()` and `.onDetach()` hooks (#882, @MichaelChirico)
 * Improved location information for R parse errors (#894, #892, @renkun-ken and @AshesITR)
+* New tag based documentation pages for linters (#888, @AshesITR)
+  * Each linter has its own help page
```
not: maybe add an example ", e.g. ?seq_linter"

MichaelChirico commented on NEWS.md:
```diff
@@ -79,6 +79,12 @@ function calls. (#850, #851, @renkun-ken)
 * Improved location information for R parse errors (#894, #892, @renkun-ken and @AshesITR)
+* New tag based documentation pages for linters (#888, @AshesITR)
+  * Each linter has its own help page
+  * `?linters` also links to tag help pages, collecting linters with a similar goal
```
idea (you may have mentioned this/similar before): it may be nice to have an API to set linters by tag in .lintr, WDYT?

AshesITR replied:
An idea might be to export alll those linter lists like `default_linters` and think of a smart way of deduplication?

e.g. make something like `linters: with_defaults(best_practices_linters, readability_linters)` work, returning all linters that are tagged any of default, best_practices or readability work?

MichaelChirico reviewed (commented):
NEWS is the only thing new to this PR right? that's all I've reviewed so far

AshesITR commented:
Sorry @MichaelChirico I saw only the last comment. Let's fix these up when preparing for the 3.0.0 release.
